### PR TITLE
Enforce label boundary when matching FQDN to proxy

### PR DESCRIPTION
**Pull request by 21KennethTran**
`ResolveFQDN` previously did not check the domain label when matching FQDN to proxy.

Fixes https://github.com/gravitational/pressure-washing/issues/415

### Test Environment
Tested locally with acme enabled, `https_keypairs` disabled, and `localhost.example.com:443` as the first option listed under `proxy_service: public_addr` on in `teleport.yaml` and this edit under `lib/service/acme.go/checkHost`
```go
defer func() {
		slog.WarnContext(ctx, "ACME host policy decision", "host", host, "allowed", err == nil, "error", err)
	}()
```
Also dynamically create app via `tctl` with the name `evillocalhost.example.com` and another with `debug-app`
### Test Cases

- [x] ACME cert issuance still works and rejects bad input
  - [x] fixed unintended behavior (see comment in pressure-washing issue)

```
// General ACME cert issuance still works

// Valid app should pass
curl -sv -k --resolve debug-app.localhost.example.com:443:127.0.0.1 https://debug-app.localhost.example.com/ 
// Nonexistent app name should fail at check
curl -sv -k --resolve nonexistent.localhost.example.com:443:127.0.0.1 https://nonexistent.localhost.example.com/ 
// Invalid suffix should fail at check
curl -sv -k --resolve debug-app.localhost.example.com.evil.com:443:127.0.0.1 https://debug-app.localhost.example.com.evil.com/

// New behavior should now reject this case (previously passes)
// Note that the correct way to call this would be evillocalhost.example.com.localhost.example.com
curl -k --resolve evillocalhost.example.com:443:127.0.0.1 https://evillocalhost.example.com/
```

### CODE REVIEW

File: lib/web/app/match.go [hunk 1]
```diff
@@ -168,6 +168,11 @@ func ResolveFQDN(
 	return nil, "", trace.NotFound("failed to resolve %v to any application within any cluster", fqdn)
 }
 
+// isHostUnderProxy tells whether host is the proxy DNS name itself or a subdomain of it.
+func isHostUnderProxy(host, proxy string) bool {
```
Comment (rhammonds-teleport) [inline lib/web/app/match.go]:
```suggestion
// hostIsProxyOrSubdomain returns true if the host is the proxy DNS name itself or if it's a subdomain of the proxy.
func hostIsProxyOrSubdomain(host, proxy string) bool {
```

File: lib/web/app/match.go [hunk 2]
```diff
@@ -134,23 +133,24 @@ func ResolveFQDN(
...
+			if !isHostUnderProxy(host, proxyMatch) {
 				return nil, "", trace.BadParameter("FQDN %q is not a subdomain of the proxy", fqdn)
 			}
 		}
 		return srv, localClusterName, nil
 	}
 
-	proxyPublicAddr := utils.FindMatchingProxyDNS(fqdn, proxyDNSNames)
-	if !strings.HasSuffix(fqdn, proxyPublicAddr) {
+	host := strings.Split(fqdn, ":")[0]
```
[the remark this answers is not in the document]
Comment (rhammonds-teleport) [reply lib/web/app/match.go]: Are you sure we should disregard this comment? At least in the ACME path, it looks we call this from `checkHost` which determines whether or not we'll retrieve new certs for a domain. Sounds like `panel.example.com:443@malicious.com` would indeed trick us into issuing a cert for the bogus FQDN.

Maybe we need a utility function that performs strict validation of FQDNs and use it in a guard clause at the top of `ResolveFQDN`.

File: lib/web/app/match.go [hunk 3]
```diff
@@ -134,23 +133,24 @@ func ResolveFQDN(
 		// suffix, so confirm the FQDN actually sits under a configured proxy DNS name before trusting a
 		// scoped match, otherwise "<hash>.somewebsite.com" resolves.
 		if srv.GetApp().GetScope() != "" {
-			host := strings.Split(fqdn, ":")[0]
 			// In the case where FindMatchingProxyDNS finds an actual found proxy match, this check is redundant.
 			// In the case where it finds no matches, FindMatchingProxyDNS returns the first element of proxyDNSNames,
 			// we must recheck to make sure that the proxy found matches, and reject if it doesn't.
+			host := strings.Split(fqdn, ":")[0]
```
Comment (zmb3) [inline lib/web/app/match.go]: I would use `net.SplitHostPort` to get the hostname, just like you did in #66890.

File: lib/web/app/match.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (21KennethTran) [reply lib/web/app/match.go]: Thanks for pointing that out. I initially thought that this path didn't see ports, and also retraced the web path which does pass in ports with `FQDNhint`. Added that utility function if you could take a second look!

File: lib/web/app/match.go [hunk 4]
```diff
@@ -168,6 +173,30 @@ func ResolveFQDN(
...
+	hostname := fqdn
+	// An error here means the fqdn isn't in proper host:port form. In that case
+	// let the IsValidHostname check below reject anything bad.
+	if h, port, err := net.SplitHostPort(fqdn); err == nil {
+		if _, err := strconv.ParseUint(port, 10, 16); err != nil {
+			return "", trace.BadParameter("invalid FQDN %q", fqdn)
+		}
+		hostname = h
+	}
+	if !utils.IsValidHostname(hostname) {
```
[the remark this answers is not in the document]
Comment (21KennethTran) [reply lib/web/app/match.go]: Interesting catch. I opted to use `IsDNS1123SubdomainWithUnderscore`, which is also used in `lib/services/app.go`.

File: lib/web/app/match.go [hunk 5]
```diff
@@ -134,23 +141,22 @@ func ResolveFQDN(
...
 		return srv, localClusterName, nil
 	}
 
 	proxyPublicAddr := utils.FindMatchingProxyDNS(fqdn, proxyDNSNames)
-	if !strings.HasSuffix(fqdn, proxyPublicAddr) {
+	if !hostIsProxyOrSubdomain(hostname, proxyPublicAddr) {
 		return nil, "", trace.BadParameter("FQDN %q is not a subdomain of the proxy", fqdn)
 	}
-	appName := strings.TrimSuffix(fqdn, fmt.Sprintf(".%s", proxyPublicAddr))
+	appName := strings.TrimSuffix(hostname, "."+proxyPublicAddr)
```
[the remark this answers is not in the document]
Comment (21KennethTran) [reply lib/web/app/match.go]: Reverted.

[the remark this answers is not in the document]
Comment (zmb3) [reply lib/web/app/match.go]: Can you add a test case that would have failed prior to reverting this change and succeeds after reverting the change?

File: lib/web/app/match.go [hunk 6]
```diff
@@ -168,8 +174,27 @@ func ResolveFQDN(
 	return nil, "", trace.NotFound("failed to resolve %v to any application within any cluster", fqdn)
 }
 
-// isHostUnderProxy tells whether host is the proxy DNS name itself or a subdomain of it.
-func isHostUnderProxy(host, proxy string) bool {
+// validateFQDN checks that fqdn is a well-formed hostname with an optional
+// numeric port and returns its hostname without the port.
+func validateFQDN(fqdn string) (string, error) {
```
Comment (zmb3) [inline lib/web/app/match.go]: Nit: I would rename this.

"validate" sounds like a function that just checks if the input is valid or not. This function actually returns the hostname.

Maybe call it `extractHostname`?

File: lib/web/app/match.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (21KennethTran) [reply lib/web/app/match.go]: Yep. Added them in the latest commit.

**Comment by zmb3** [comment]
This one is missing a test plan.

The finding says:

> The net effect is that a user with app:create rights in any cluster (e.g., a leaf admin) can drive the root proxy's autocert.Manager to request — and persistently cache — Let's Encrypt certificates for hostnames the proxy operator never put in proxy_service.public_addr.

So I would expect a test plan that:

1. Makes sure ACME cert issuance still works
2. Makes sure that the behavior mentioned above is no longer allowed.

**Review by rhammonds-teleport — APPROVED** [review]
Looks to me like we've got all the edge cases ironed out. 👍 

Would it be feasible to set up a test cluster to exercise some of these app access scenarios and add them to the manual test plan?

**Comment by zmb3** [comment]
What did your `teleport.yaml` look like for the manual test plan that you ran?

It looks like you are using localhost and I'm not sure you'd be able to solve ACME challenges running locally.

**Comment by 21KennethTran** [comment]
> What did your `teleport.yaml` look like for the manual test plan that you ran?
> 
> It looks like you are using localhost and I'm not sure you'd be able to solve ACME challenges running locally.

I enabled acme, commented out the `https_keypairs` section, and added this to my `public_addr` section to get this to work.

```
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr:
    - localhost.example.com:443
    - localhost:443
    - 127.0.0.1:443
  # https_keypairs:   # ignore when acme on
  # - key_file: /Users/kennethtran/teleport-test-env/teleport-tls/localhost-key.pem
  #   cert_file: /Users/kennethtran/teleport-test-env/teleport-tls/localhost.pem
  https_keypairs_reload_interval: 0s
  acme:
    enabled: "yes"
    uri: https://acme-staging-v02.api.letsencrypt.org/directory   # Let's encrypt staging
```
